scrapy/core/downloader/handlers/http11.py

```diff
@@ -122,7 +122,15 @@ def processProxyResponse(self, bytes):
         """
         self._protocol.dataReceived = self._protocolDataReceived
         if  TunnelingTCP4ClientEndpoint._responseMatcher.match(bytes):
-            self._protocol.transport.startTLS(self._contextFactory,
+            try:
+                # this sets proper Server Name Indication extension
+                # but is only available for Twisted>=14.0
+                sslOptions = self._contextFactory.creatorForNetloc(
+                    self._tunneledHost, self._tunneledPort)
+            except AttributeError:
+                # fall back to non-SNI SSL context factory
+                sslOptions = self._contextFactory
```

Comment:
@kmike , I don't like `connectionCreator` better tbh, sounds too generic, not related to securing the already opened transport link.
Even if [the interface says](https://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.ITLSTransport.startTLS.html) `IOpenSSLClientConnectionCreator` (in our case), it also says:

> For clients, use `twisted.internet.ssl.optionsForClientTLS`;
> for servers, use `twisted.internet.ssl.CertificateOptions`.

Twisted's examples [also use "options"](https://twistedmatrix.com/documents/current/core/howto/ssl.html#using-starttls) in some way.

It could be `tlsOptions` though.
Or `tlsContextFactory` to keep the "factory" idiom.